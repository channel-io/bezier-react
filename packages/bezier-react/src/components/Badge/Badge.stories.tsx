import React from 'react'

import { AppleIcon } from '@channel.io/bezier-icons'
import { type Meta, type StoryFn, type StoryObj } from '@storybook/react'

import { Badge } from './Badge'
import { type BadgeProps } from './Badge.types'

const meta: Meta<typeof Badge> = {
  component: Badge,
}

export default meta

const Template: StoryFn<BadgeProps> = ({ children, ...badgeProps }) => (
  <Badge {...badgeProps}>{children}</Badge>
)

export const Primary: StoryObj<BadgeProps> = {
  render: Template,

  args: {
    children: 'Design',
    size: 'm',
    icon: AppleIcon,
    variant: 'default',
  },
}
