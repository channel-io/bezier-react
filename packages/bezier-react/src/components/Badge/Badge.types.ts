import { type BezierIcon } from '@channel.io/bezier-icons'

import {
  type AlphaBezierComponentProps,
  type ChildrenProps,
  type SizeProps,
  type VariantProps,
} from '~/src/types/props'

import {
  type TagBadgeSize,
  type TagBadgeVariant,
} from '~/src/components/TagBadgeCommon'

interface BadgeOwnProps {
  /**
   * Icon to be shown on the left side of the badge.
   */
  icon?: BezierIcon
}

interface BadgeProps extends
  AlphaBezierComponentProps,
  ChildrenProps,
  Omit<React.HTMLAttributes<HTMLDivElement>, 'color'>,
  SizeProps<TagBadgeSize>,
  VariantProps<TagBadgeVariant>,
  BadgeOwnProps {}

export default BadgeProps
