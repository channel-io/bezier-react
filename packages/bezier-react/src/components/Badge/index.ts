export { Badge } from './Badge'
export type { BadgeProps, BadgeSize, BadgeVariant } from './Badge.types'
