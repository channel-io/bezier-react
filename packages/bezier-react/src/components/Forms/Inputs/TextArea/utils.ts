import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
