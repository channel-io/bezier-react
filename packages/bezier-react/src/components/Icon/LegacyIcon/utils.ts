/* Internal Dependencies */
// eslint-disable-next-line no-restricted-imports
import icons, { type IconName } from '../generated'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { Icon, AllIcon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-react'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(<svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
