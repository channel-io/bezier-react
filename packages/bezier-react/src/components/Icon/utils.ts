import { type SVGProps } from 'react'

const BEZIER_ICON_SYMBOL = Symbol('bezier.icon')

export type IconSource = (props: SVGProps<SVGSVGElement>) => JSX.Element

export type BezierIcon = IconSource & { [BEZIER_ICON_SYMBOL]: true }

export function isBezierIcon(arg: unknown): arg is BezierIcon {
  return (
    typeof arg === 'function' &&
    arg[BEZIER_ICON_SYMBOL] === true
  )
}

export function createBezierIcon(source: IconSource): BezierIcon {
  source[BEZIER_ICON_SYMBOL] = true
  return source as BezierIcon
}
