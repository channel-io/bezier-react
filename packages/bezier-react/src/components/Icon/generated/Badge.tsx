import * as React from 'react'
import { SVGProps } from 'react'
import { createBezierIcon } from '../utils'
function SvgBadge(props: SVGProps<SVGSVGElement>) {
  return (
    <svg
      width="1em"
      height="1em"
      viewBox="0 0 24 24"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}
    >
      <path
        fill="currentColor"
        fillRule="evenodd"
        clipRule="evenodd"
        d="M12 3a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13Zm4 14.002a8.5 8.5 0 1 0-8 0v5.16a.5.5 0 0 0 .737.441L12 20.846l3.263 1.757a.5.5 0 0 0 .737-.44v-5.161Zm-.303-9.72a1 1 0 0 1 .02 1.414l-3.882 4a1 1 0 0 1-1.435 0l-2.118-2.181a1 1 0 0 1 1.436-1.393l1.4 1.442 3.164-3.26a1 1 0 0 1 1.415-.022Z"
      />
    </svg>
  )
}
export default createBezierIcon(SvgBadge)
