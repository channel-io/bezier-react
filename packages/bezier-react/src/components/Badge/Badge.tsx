import React, {
  forwardRef,
  memo,
} from 'react'

import classNames from 'classnames'

import { isEmpty } from '~/src/utils/type'

import { Icon } from '~/src/components/Icon'
import { getProperTagBadgeTypo } from '~/src/components/TagBadgeCommon'
import commonStyles from '~/src/components/TagBadgeCommon/TagBadge.module.scss'
import { Text } from '~/src/components/Text'

import { type BadgeProps } from './Badge.types'

import styles from './Badge.module.scss'

export const BADGE_TEST_ID = 'bezier-badge'

/**
 * `Badge` is a component for representing badge, which consists of text and icon.
 * @example
 * ```tsx
 * <Badge
 *   size="xs"
 *   variant="blue"
 *   icon={AppleIcon}
 * >
 *   Beta
 * </Badge>
 * ```
 */
export const Badge = memo(forwardRef<HTMLDivElement, BadgeProps>(function Badge({
  size = 'm',
  variant = 'default',
  icon,
  children,
  className,
  ...rest
}, forwardedRef) {
  return (
    <div
      ref={forwardedRef}
      className={classNames(
        styles.Badge,
        styles[`variant-${variant}`],
        commonStyles.TagBadge,
        commonStyles[`size-${size}`],
        className,
      )}
      data-testid={BADGE_TEST_ID}
      {...rest}
    >
      { icon && (
        <Icon
          source={icon}
          size="xs"
        />
      ) }

      { !isEmpty(children) && (
        <Text
          typo={getProperTagBadgeTypo(size)}
          marginHorizontal={3}
        >
          { children }
        </Text>
      ) }
    </div>
  )
}))
