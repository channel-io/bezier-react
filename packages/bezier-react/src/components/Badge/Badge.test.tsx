import React from 'react'

import { render } from '~/src/utils/test'

import { BADGE_TEST_ID, Badge } from './Badge'
import { type BadgeProps } from './Badge.types'

describe('Badge test >', () => {
  let props: BadgeProps

  beforeEach(() => {
    props = {}
  })

  const renderBadge = (optionProps?: BadgeProps) =>
    render(
      <Badge
        {...props}
        {...optionProps}
      />
    )

  it('Snapshot >', () => {
    const { getByTestId } = renderBadge()
    const badge = getByTestId(BADGE_TEST_ID)

    expect(badge).toMatchSnapshot()
  })
})
