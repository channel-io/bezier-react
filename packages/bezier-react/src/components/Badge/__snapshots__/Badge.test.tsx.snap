// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge test > Snapshot > 1`] = `
<div
  class="Badge variant-Default TagBadge size-m"
  data-testid="bezier-badge"
/>
`;
